codegen/main.go

```diff
@@ -0,0 +1,461 @@
+	"flag"
+	"fmt"
+	"os"
+	"regexp"
+	"slices"
+	"strings"
+	"text/template"
+	"time"
+	"unicode"
+
+	"golang.org/x/exp/maps"
+
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	outF  = flag.String("outF", "", "Output file")
+	inF   = flag.String("inF", "", "Input file")
+	tmplF = flag.String("tmplF", "", "Template file")
+
+	paramNameRegexCompiled = regexp.MustCompile(`[a-zA-Z][.a-zA-Z0-9\-]*`)
+)
+
+// Represents the fields Config structs.
+type field struct {
+	Name       string
+	DataType   string
+	ConfigPath string
+}
+type typeTemplateData struct {
+	// Name of the type
+	Name string
+	// Fields that are to be included in the
+	Fields []field
+}
+
+type flagTemplateData struct {
+	Name               string
+	Shorthand          string
+	Type               string
+	DefaultValue       string
+	ConfigPath         string
+	Fn                 string
+	Usage              string
+	IsDeprecated       bool
+	DeprecationWarning string
+	IsHidden           bool
+	IsShorthandHidden  bool
+}
+
+type templateData struct {
+	TypeTemplateData []typeTemplateData
+	FlagTemplateData []flagTemplateData
+	// Back-ticks are not supported in templates. So, passing as a parameter.
+	BT string
+}
+
+type Param struct {
+	FlagName           string `yaml:"flag-name"`
+	Shorthand          string
+	Type               string
+	DefaultValue       string `yaml:"default"`
+	ConfigPath         string `yaml:"config-path"`
+	IsDeprecated       bool   `yaml:"deprecated"`
+	DeprecationWarning string `yaml:"deprecation-warning"`
+	Usage              string
+	HideFlag           bool `yaml:"hide-flag"`
+	HideShorthand      bool `yaml:"hide-shorthand"`
+}
+
+func capitalizeIdentifier(name string) (string, error) {
+	if !paramNameRegexCompiled.MatchString(name) {
+		return "", fmt.Errorf("%s is not a supported name", name)
+	}
+	id := ""
+	capitalizeNextRune := true
+	for _, v := range name {
+		if v != '-' && v != '.' {
+			if capitalizeNextRune {
+				v = unicode.ToUpper(v)
+			}
+			id = id + string(v)
+			capitalizeNextRune = false
+		} else {
+			capitalizeNextRune = true
+		}
+	}
+	return id, nil
+}
+
+type typeInfo struct {
+	typeName   string
+	fieldName  string
+	configPath string
+	dataType   string
+}
+
+func getGoDataType(dt string) (string, error) {
+	switch dt {
+	case "octal":
+		return "Octal", nil
+	case "duration":
+		return "time.Duration", nil
+	default:
+		return dt, nil
+	}
+}
+
+// Figures out what all structs and corresponding fields need to be created for
+// a given param. Other params can add fields to the same type if they share the
+// same config-path hierarchy.
+// For a config-path of x.y for an int param it will yield the following structs
+//
+//	type Config struct {
+//	  X XConfig `yaml:"x"`
+//	}
+//
+//	type XConfig struct {
+//	  Y int `yaml:"y"`
+//	}
+func typesToCreate(param Param) ([]typeInfo, error) {
+	configPath := param.ConfigPath
+	typeNameSuffix := "Config"
+	typesFieldPairs := make([]typeInfo, 0)
+	for configPath != "" {
+		nestedTypeName := ""
```

Comment:
Nit: You can alternatively just declare string variables (by default they will be empty string)